src/shared/telemetry/defaultTelemetryService.ts

```diff
@@ -262,6 +271,32 @@ export class DefaultTelemetryService implements TelemetryService {
             return []
         }
     }
+
+    /**
+     * Only passive telemetry is allowed during startup (except for some known
+     * special-cases).
+     */
+    public assertPassiveTelemetry(didReload: boolean) {
+        // Special case: these may be non-passive during a VSCode "reload". #1592
+        const activeAllowed = ['sam_init']
+        // Metrics from the previous session can be arbitrary: we can't reason
+        // about whether they should be passive/active.
+        let readingCache = true
+
+        // Array for..of is in-order:
+        // https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/forEach
+        for (const metric of this._eventQueue) {
+            if (readingCache) {
+                readingCache = metric !== this._endOfCache
+                // Skip cached metrics.
+                continue
+            }
+            if (metric.Passive || (didReload && activeAllowed.includes(metric.MetricName))) {
+                continue
+            }
+            throw Error(`non-passive metric emitted at startup: ${metric.MetricName}`)
+        }
+    }
```

Comment:
Can this for loop be replaced with just the final if statement and look only at `this._endOfCache`?

Reply:
No... `this._endOfCache` marks the last item before the non-cached items. All non-cached items need to be checked on startup (i.e. any telemetry that was queued during activation).